File: #200_Number_of_Islands/Solution.java
import java.util.*;

public class Solution {

	int rows;
	int cols;

    public int numIslands(char[][] grid) {
        int result = 0;
        if (grid == null || grid.length == 0 || grid[0].length == 0) return result;
        rows = grid.length;
        cols = grid[0].length;
        boolean[][] isMarked = new boolean[rows][cols];
        for (int i = 0; i < rows; i++) {
        	for (int j = 0; j < cols; j++) {
        		if (grid[i][j] == '1' && !isMarked[i][j]) {
        			result++;
        			explore(grid, isMarked, i, j);
        		}
        	}
        }
        return result;
    }

    public void explore(char[][] grid, boolean[][] isMarked, int i, int j) {
    	isMarked[i][j] = true;
    	if (i - 1 >= 0 && grid[i - 1][j] == '1' && !isMarked[i - 1][j])
    		explore(grid, isMarked, i - 1, j);
    	if (j - 1 >= 0 && grid[i][j - 1] == '1' && !isMarked[i][j - 1])
    		explore(grid, isMarked, i, j - 1);
    	if (i + 1 < rows && grid[i + 1][j] == '1' && !isMarked[i + 1][j])
    		explore(grid, isMarked, i + 1, j);
    	if (j + 1 < cols && grid[i][j + 1] == '1' && !isMarked[i][j + 1])
    		explore(grid, isMarked, i, j + 1);
    }

    public static void main(String[] args) {
    	char[][] grid = {
    		{'1', '1', '0', '0', '0'},
    		{'1', '1', '0', '0', '0'},
    		{'0', '0', '1', '0', '0'},
    		{'0', '0', '0', '1', '1'}
    	};
    	Solution sol = new Solution();
    	System.out.println(sol.numIslands(grid));
    }
}

File: #200_Number_of_Islands/Solution.py
class Solution(object):

	def __init__(self):
		self.count = 0
		self.visited = None

	def numIslands(self, grid):
		if not grid or not grid[0]:
			return 0
		self.count = 0
		self.visited = [[False for j in range(len(grid[0]))] for i in range(len(grid))]
		for i in range(len(grid)):
			for j in range(len(grid[0])):
				if grid[i][j] == '0':
					continue
				if grid[i][j] == '1' and self.visited[i][j]:
					continue
				self.dfs(grid, i, j)
				self.count += 1
		return self.count

	def dfs(self, grid, i, j):
		if i < 0 or i >= len(grid):
			return
		if j < 0 or j >= len(grid[0]):
			return
		if self.visited[i][j] or grid[i][j] == '0':
			return
		self.visited[i][j] = True
		self.dfs(grid, i - 1, j)
		self.dfs(grid, i + 1, j)
		self.dfs(grid, i, j - 1)
		self.dfs(grid, i, j + 1)

sol = Solution()
t = [['1', '1', '0', '0', '0'], 
	['1', '1', '0', '0', '0'],
	['0', '0', '1', '0', '0'],
	['0', '0', '0', '1', '1']]
print sol.numIslands(t)